pkg/resto/resto.go

```diff
@@ -0,0 +1,326 @@
+		return nil, err
+	}
+	if mediaType == MediaTypeConfig {
+		var ma ManifestAny
+		err = json.Unmarshal(payload, &ma)
+		if err != nil {
+			return nil, err
+		}
+		res := make(map[string]string)
+		err = json.Unmarshal([]byte(ma.Payload), &res)
+		return res, err
+	}
+	// legacy image mode
+	refs := manifest.References()
+	if len(refs) != 2 {
+		return nil, fmt.Errorf("expected 2 references, found %v", len(refs))
+	}
+	// assume second element is the layer (first being the image config)
+	r := refs[1]
+	rdgst := r.Digest
+	blobsService := repo.Blobs(ctx)
+	payloadGz, err := blobsService.Get(ctx, rdgst)
+	if err != nil {
+		return nil, err
+	}
+	payloadBuf := bytes.NewBuffer(payloadGz)
+	gzf, err := gzip.NewReader(payloadBuf)
+	if err != nil {
+		return nil, err
+	}
+	tarReader := tar.NewReader(gzf)
+	return tarContent(tarReader)
+}
+
+func tarContent(tarReader *tar.Reader) (map[string]string, error) {
+	res := make(map[string]string)
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if header.Typeflag == tar.TypeReg {
+			content := bytes.NewBuffer(nil)
+			io.Copy(content, tarReader)
+			res[header.Name] = content.String()
+		}
+	}
+	return res, nil
+}
+
+// PushConfig pushes a configuration file to a registry and returns its digest
+func PushConfig(ctx context.Context, payload string, repoTag string, username, password string, insecure bool, labels map[string]string) (string, error) {
+	return PushConfigMulti(ctx, map[string]string{
+		"config": payload,
+	}, repoTag, username, password, insecure, labels)
+}
+
+// PushConfigMulti pushes a set of configuration files to a registry and returns its digest
+func PushConfigMulti(ctx context.Context, payload map[string]string, repoTag string, username, password string, insecure bool, labels map[string]string) (string, error) {
+	pr, err := parseRef(repoTag)
+	if err != nil {
+		return "", err
+	}
+	if username == "" {
+		username, password = getCredentials(pr.domain)
+	}
+	j, err := json.Marshal(payload)
+	if err != nil {
+		return "", err
+	}
+	manifestAny := ManifestAny{
+		Versioned: manifest.Versioned{
+			SchemaVersion: 2,
+			MediaType:     MediaTypeConfig,
+		},
+		Payload: string(j),
+	}
+	raw, err := json.Marshal(manifestAny)
+	if err != nil {
+		return "", err
+	}
+	repo, err := NewRepository(ctx, pr.domain, pr.path, username, password, insecure)
+	if err != nil {
+		repo, err = NewRepository(ctx, strings.Replace(pr.domain, "https://", "http://", 1), pr.path, username, password, insecure)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	manifestService, err := repo.Manifests(ctx)
+	if err != nil {
+		return "", err
+	}
+	manifest := NewConfigManifest(MediaTypeConfig, raw)
+	dgst, err := manifestService.Put(ctx, manifest, distribution.WithTag(pr.tag))
+	if err == nil {
+		return dgst.String(), nil
+	}
+	if !strings.Contains(err.Error(), "manifest invalid") && !strings.Contains(err.Error(), "manifest Unknown") {
+		return "", err
+	}
+	// try legacy mode
```

Comment:
Would be good to separate out the legacy and new modes into functions so that this is easier to follow